### ISSUE TITLE
core: Python 3 support

### ISSUE THREAD

**Issue by dw** [dw/mitogen#16]
Too many changes to even consider just yet

**Comment by href** [comment]
Any chance of this being something that could be done in the light of your announced Kickstarter project? I'd love to integrate mitogen into [Suitable](http://suitable.readthedocs.io) but I'm just not willing to go back to Python 2 :)

**Comment by dw** [comment]
We have at least simple, solid plan, but so far I don't like it -- it would mean 3.x masters can't talk to 2.x children. Still a work in progress, I want to fix this without dropping 2.4 support. Ultimately it's probably simple work, but the _testing_ involved is why I can't include it in the initial 3 weeks, which are already a stretch simply due to the testing work. But if you back the project I'll treat this as a priority and try squeeze it in for the 1 month mark :)

**Comment by href** [comment]
Deal. Time to convince my boss to send some money your way :)

**Comment by dw** [comment]
You champion!

**Comment by href** [comment]
Done. Seantis, my employer, should have made a pledge :)

**Comment by dw** [comment]
That is beyond fabulous! Please thank him in person for me, will be in
touch for logo and suchlike once the campaign is finished.

Regarding your desire for 3.x support, is your environment pure 3.x-only,
or is it for example, 3.x masters talking to 2.x targets?

I had completely forgotten during our original discussion that preliminary
3.x is already partly complete, using the '2to3' method in setup.py. Tthis
allows for 3.x to 3.x communication, but not 3.x to 2.x, which I still want
to implement. However, this all needs thorough testing, and most certainly
I haven't been tested it much even during development.

**Comment by href** [comment]
Our environment is mixed. Though 3.x to 3.x is probably okay, as long as we can specify the path to the Python binary.

I’m mainly interested in speeding up our Suitable code base and I need to support Python 2/3 there. So I expect we’ll use our support hours to have you help us out there :)

**Comment by dw** [comment]
Just to be clear, I definitely don't want to settle on just 3.x->3.x. :)
If Ansible can support 3.x->2.x, then Mitogen needs to handle it properly,
otherwise it's not doing its job.

Thanks again so much

**Comment by href** [comment]
You are welcome :)

**Comment by dw** [comment]
Chatting about 3.x support, the Ansible guys pointed me to https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/pycompat24.py#L32 .. which solves the single major issue with having a single unified 2.x/3.x codebase that remains compatible all the way back to 2.4.

This hack only needs to happen in 8 places in the code, looks like we've finally got a solid plan for 2.x/3.x support :)

**Comment by moreati** [comment]
Some 2.x `cPickle`/`pickle` and 3.x `_pickle`/`pickle` differences I stumbled on, and wanted to make a note of

1. In Python 2.x [Subclassing Unpicklers](https://docs.python.org/2/library/pickle.html#subclassing-unpicklers)
    1. `cPickle.Unpickler` requires replacing `Unpickler.load_global(mod_name, class_name)`
    1. `pickle.Unpickler` requires deriving a subclass that overrides `Unpickler.load_global()` with a method that does imports and fiddles with the unpickler stack
        > If this sounds like a hack, you're right. Refer to the source code to make this work.
1. In Python 3.x [Restricting Globals](https://docs.python.org/3/library/pickle.html#restricting-globals)
    1. `pickle.Unpickler` may be a pure Python implementation, or a C implementation from `_pickle`
    1.  the method to override was _renamed_ to `find_class(mod_name, name)` with the semantics of the old `cPickle.Unpickler`

**Comment by dw** [comment]
One more pickle difference: for protocol=2, Python 3 parses 2.x `str` as Unicode, and expects `bytes` to appear as a function call to `codecs.encode(..., "latin1")`. This hack allows 3.x to decode class instances from 2.x, where in 3.x instance attribute names must be Unicode, whereas in 2.x they were str.

As a general update on this, `python3` branch is working relatively well already, including 2.x->3.x and 3.x->2.x. There is a bunch more work remaining, including a change to how module dependency scanning works. The bytecode format has changed at some point on 3.x, so we need tweaks to catch up.

Hoping to have the branch in a mergeable state sometime end of May.

**Comment by moreati** [comment]
One more 2.x vs 3.x difference with pickle. Python 3.x encodes integers outside the range `-2**31 <= i <= 2**31-1` using opcodes for `long()` type. 64 bit builds of Python 2.x encode these integers using opcodes for `int()`.

So the value 4294967296 (`2**32`) does not roundtrip from Python 2.x -> 3.x -> 2.x. It comes back out as 4294967296**L**. This probably an obscure quirk that can be ignored, but I wanted to document it.

**Comment by moreati** [comment]
> 4294967296 (2**32) does not roundtrip from Python 2.x -> 3.x -> 2.x. It comes back out as 4294967296L
On further thought, this is probably also true of Python 2.x 64-bit -> 2.x 32-bit -> 2.x 64-bit.

**Comment by dw** [comment]
Python 3 Ansible host targetting Python 2.x child is now up to 173 steps of the integration tests. :)

I wouldn't even try it against a 3.x child yet -- there is definitely more work required. Meanwhile for 2.x targets, it is likely already working well for simpler playbooks. This shouldn't take much longer. Sorry for the delay!

**Comment by willmerae** [comment]
Victor Stinner (a core dev) spoke at PyLondinum yesterday, about the history of Python 3.x. Two slides caught my eye, as quirks/behaviour I didn't know about. Summarized below

> Python 2.7 WONTFIX
> - ...
> - subprocess is not thread safe
> - threading.RLock is not signal safe
> - ...

> Fixed in Python 3
> - ...
> - 3.4: file descriptors non-inheritable, fork+exec safety (PEP 446)
> - 3.5: retry syscalls on EINTR (PEP 475)

**Comment by RonnyPfannschmidt** [comment]
as a note - execnet dropped pickle and implemented a own serializer to bridge the gap between py2 and py3

**Comment by dw** [comment]
Hi Ronny :)

There are enough hacks to make pickle work in python3 branch -- the commits displayed in the GitHub UI no longer match the branch contents, as I've clearly been rebasing like crazy, and decided to just remove the issue references to stop spamming the ticket.

The branch is almost ready to merge - the built in tests pass for 2->3, 3->2, 3->3 and 2->2, however a big hang has been introduced that only manifests under the Ansible tests, and there is at least one hang on master. I need to find and fix both hangs before merging the result together.

Pickle definitely needs to go -- https://github.com/dw/mitogen/issues/126 is open for that, but it needs to stay at least for the first stable series, as that is now somewhere around 8 weeks behind schedule ;(

**Comment by RonnyPfannschmidt** [comment]
so for now i wont touch it ^^

**Comment by dw** [comment]
This is now (finally) on master. A couple of inert races left that impact the unit tests around child process reaping, but not Ansible itself, so not delaying this any longer.
